src/Storages/NATS/WriteBufferToNATSProducer.cpp

```diff
@@ -0,0 +1,174 @@
+    std::optional<char> delimiter,
+    size_t rows_per_message,
+    size_t chunk_size_)
+    : WriteBuffer(nullptr, 0)
+    , connection(configuration_, log_)
+    , subject(subject_)
+    , shutdown_called(shutdown_called_)
+    , payloads(BATCH)
+    , log(log_)
+    , delim(delimiter)
+    , max_rows(rows_per_message)
+    , chunk_size(chunk_size_)
+{
+    if (!connection.connect())
+        throw Exception(ErrorCodes::CANNOT_CONNECT_NATS, "Cannot connect to NATS {}", connection.connectionInfoForLog());
+
+    writing_task = global_context->getSchedulePool().createTask("NATSWritingTask", [this] { writingFunc(); });
+    writing_task->deactivate();
+
+    reinitializeChunks();
+}
+
+
+WriteBufferToNATSProducer::~WriteBufferToNATSProducer()
+{
+    writing_task->deactivate();
+    connection.disconnect();
+    assert(rows == 0);
+}
+
+
+void WriteBufferToNATSProducer::countRow()
+{
+    if (++rows % max_rows == 0)
+    {
+        const std::string & last_chunk = chunks.back();
+        size_t last_chunk_size = offset();
+
+        if (last_chunk_size && delim && last_chunk[last_chunk_size - 1] == delim)
+            --last_chunk_size;
+
+        std::string payload;
+        payload.reserve((chunks.size() - 1) * chunk_size + last_chunk_size);
+
+        for (auto i = chunks.begin(), end = --chunks.end(); i != end; ++i)
+            payload.append(*i);
+
+        payload.append(last_chunk, 0, last_chunk_size);
+
+        reinitializeChunks();
+
+        ++payload_counter;
+        if (!payloads.push(payload))
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Could not push to payloads queue");
+    }
+}
+
+void WriteBufferToNATSProducer::publish()
+{
+    uv_thread_t flush_thread;
+
+    uv_thread_create(&flush_thread, publishThreadFunc, static_cast<void *>(this));
+
+    connection.getHandler().startLoop();
+    uv_thread_join(&flush_thread);
+}
+
+void WriteBufferToNATSProducer::publishThreadFunc(void * arg)
+{
+    String payload;
+    WriteBufferToNATSProducer * buffer = static_cast<WriteBufferToNATSProducer *>(arg);
+
+    natsStatus status;
+    while (!buffer->payloads.empty())
+    {
+        bool pop_result = buffer->payloads.pop(payload);
+
+        if (!pop_result)
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Could not pop payload");
+        status = natsConnection_PublishString(buffer->connection.getConnection(), buffer->subject.c_str(), payload.c_str());
+
+        if (status != NATS_OK)
+        {
+            LOG_DEBUG(buffer->log, "Something went wrong during publishing to NATS subject. Nats status text: {}. Last error message: {}",
+                      natsStatus_GetText(status), nats_GetLastError(nullptr));
+            if (!buffer->payloads.push(std::move(payload)))
+                throw Exception(ErrorCodes::LOGICAL_ERROR, "Could not push to payloads queue");
+            break;
+        }
+    }
+
+    nats_ReleaseThreadMemory();
+}
+
+
+void WriteBufferToNATSProducer::writingFunc()
```

Comment:
let's add try catch with tryLogCurrentException in catch, afaik if we do not do that for task here, it will not be printed at all